File: src/main/clojure/au/com/permeance/clojure/clojure_portlet.clj
(ns au.com.permeance.clojure.clojure-portlet
  (:use au.com.permeance.clojure.portlet))

(gen-class
  :name au.com.permeance.clojure.ClojurePortlet
        :extends javax.portlet.GenericPortlet
        :main false
        :methods [[^{javax.portlet.ProcessAction {:name "clojure-action"}}
                   processClojureAction [javax.portlet.ActionRequest javax.portlet.ActionResponse] void]

                  [^{javax.portlet.RenderMode {:name "view"}}
                   processClojureRender [javax.portlet.RenderRequest javax.portlet.RenderResponse] void]])

(defn -processClojureAction [portlet request response]
  (set-render-parameters response {:message "An action occurred"}))

(defn -processClojureRender [portlet request response]
  (let [param (first (:message (get-request-parameters request)))
        message (or param "Nothing happened")]
    (dispatch portlet "/view.jsp" {:message message} request response)))

File: src/main/java/au/com/permeance/clojure/ClojurePortlet.java
package au.com.permeance.clojure;

import javax.portlet.ActionRequest;
import javax.portlet.ActionResponse;
import javax.portlet.GenericPortlet;
import javax.portlet.PortletContext;
import javax.portlet.PortletException;
import javax.portlet.PortletRequestDispatcher;
import javax.portlet.ProcessAction;
import javax.portlet.RenderMode;
import javax.portlet.RenderRequest;
import javax.portlet.RenderResponse;
import java.io.IOException;

public class ClojurePortlet extends GenericPortlet {

    @ProcessAction(name = "clojure-action")
    public void processClojureAction(ActionRequest request, ActionResponse response) {
        response.setRenderParameter("message", "An action occurred");
    }

    @RenderMode(name = "view")
    public void processClojureRender(RenderRequest request, RenderResponse response) throws PortletException, IOException {
        String message = request.getParameter("message");
        if (message == null) {
            message = "Nothing happened";
        }
        request.setAttribute("message", message);

        PortletContext context = getPortletContext();
        PortletRequestDispatcher dispatcher = context.getRequestDispatcher("/view.jsp");
        dispatcher.include(request, response);
    }

}
